### DISCUSSION TITLE
Assembly language programming

### DISCUSSION THREAD

**Discussion by ghhann** [picocomputer/community#127] [Q&A]
Been reviewing all get it started and development videos in YouTube. At one point it was mention that a video on assembly language programming was going to be made.   I was looking for it does anyone know the link

All my best to everyone

**Comment by varomix** [comment]
I’m also interested in Assembly programming for the Picocomputer, thanks

**Comment by GlennFR** [comment]
Hi ghhann, varomix.  Can you be more precise as to what you want to know/learn.  There is a whole host of 6502 assembly programming on many forums, most notably [forum.6502.org](http://forum.6502.org/index.php)
If you are already up-to-speed with the assembly programming, but want help in interfacing to the RIA, then I could provide some help/pointers.

**Comment by ghhann** [comment]
Thank you for your response.  I was looking on how to setup the tools to interface with the ria.  Maybe a step by step guide would be most appreciated with a simple hello world.   I am stuck in setting up compiling llvm mos.  I was able to get the core llvm installed and compiled so I have clang installed. I think it would be a benefit to the community to have a step by step guide on paper.   Sort of the old commodore programmers reference guide obviously with out the nemonics details Sorry if it does not make sense have bad cte in my head BestOn Oct 30, 2024, at 12:38 PM, GlennFR ***@***.***> wrote:﻿
Hi ghhann, varomix.  Can you be more precise as to what you want to know/learn.  There is a whole host of 6502 assembly programming on many forums, most notably forum.6502.org
If you are already up-to-speed with the assembly programming, but want help in interfacing to the RIA, then I could provide some help/pointers.

**Comment by rumbledethumps** [comment]
6502 assembly on LLVM-MOS isn't as nice as CC65. The CC65 library is a good place to start (see ria.s).
https://github.com/picocomputer/cc65/tree/master/libsrc/rp6502

**Comment by GlennFR** [comment]
I second Rumble's comment about the cc65 assembler.  I have some code snippets (actually written for my development of the PLASMA language for the rp6502 - which, unfortunately, is on the back-burner for the moment : real work is more important).  Snippets include reading/writing to the screen, some basic file I/O.   You can see how the stack is used, etc.   Is this what you're looking for?

**Comment by ghhann** [comment]
Yes I’ll give it a try I’ve used and compiled cc65 for some c64 stuff.   I’ll give it a try for the pico computer.  Best to everyone On Oct 31, 2024, at 5:58 AM, GlennFR ***@***.***> wrote:﻿
I second Rumble's comment about the cc65 assembler.  I have some code snippets (actually written for my development of the PLASMA language for the rp6502 - which, unfortunately, is on the back-burner for the moment : real work is more important).  Snippets include using reading/writing to the screen, some basic file I/O.   You can see how the stack is used, etc.   Is this what you're looking for?

**Comment by grakoczy** [comment]
Hi,
I've built Ben Eater's breadboard computer based on 6502 and played a little bit with assebly code. 
I would love to see some template project for assembler similar like we already have for cc65 or llvm-mos. 
If it could contain some examples for RIA and VGA it would be really great. 
As there are a lot of resources where to learn 6502 assembly language all we need it is some examples and template to begin.

**Comment by ghhann** [comment]
When I was working in IT (I am retired ) I learned 6502 addressing from a book from a magazine called Compute! Was machine language for beginners a pdf can be obtained through google had a good chapter 6 that went into the instruction set. What would help me is a breakdown of the memory map and kernel routines that I can explore and a graphics, sound primer.  One of the things I enjoyed was the old commodore programmers reference guides that were published for the 64 and 128 and even the Amiga kernel disassembly books Fyi the only reason I mention my age was that I lived during a great early computer period that I need to do z80a programming on cpm operating system “mainframe” terminal period a time that some people are trying to recreate as a retro period. I still have some of my loader programs on punchtape and timeshare logs on program execution time on cpu for department billing’s Best On Oct 31, 2024, at 6:44 AM, grakoczy ***@***.***> wrote:﻿
Hi,
I've built Ben Eater's breadboard computer based on 6502 and played a little bit with assebly code.
I would love to see some template project for assembler similar like we already have for cc65 or llvm-mos.
If it could contain some examples for RIA and VGA it would be really great.
As there are a lot of resources where to learn 6502 assembly language all we need it is some examples and template to begin.

**Reply by netzerohero** [reply]
Re:
> What would help me is a breakdown of the memory map and kernel routines that I can explore and a graphics, sound primer.

Since you mentioned "memory map", maybe this thread from Dec-2023 would be of interest:
https://github.com/orgs/picocomputer/discussions/64

**Comment by vruumllc** [comment]
One thing that I would like to see is a project example using mixed C and assembly files.

**Reply by netzerohero** [reply]
It may not be the best example... but for an example of mixed assembly & C, take a look in the ehbasic-plus repo:
[ehbasic-plus](https://github.com/picocomputer/ehbasic-plus/tree/main)/[src](https://github.com/picocomputer/ehbasic-plus/tree/main/src)

Lee Davison's ehbasic source is entirely 6502 assembly.
 
In file:
rp6502enhance.s
... I call C-code from assembly in the CC65 environment.
That's where I mixed some of your C-graphics-library code with the ehbasic-language assembly source - to provide initial graphics support for Basic. 

One will have to read-up on the calling conventions when mixing assembly & C on the 6502:
https://cc65.github.io/doc/cc65-intern.html#ss2.1

**Reply by patrick-flynn** [reply]
@vruumllc I wrote a demo that vertically scrolls a 290x577 image through a 320x240x4-bit canvas. The image is bigger than XRAM, so I read a screen sized chunk from a file to initialize the display, and single-line chunks to fill in a row when it's uncovered by scrolling (so one `seek()` and then one `read_xram()` of 290/2 = 145 bytes is all that is needed per scrolling event). When I implemented the scrolling in C, it took longer than a jiffy, so I reimplemented it in assembly - either direction; two tight loops. That runs shorter than a jiffy -- now, the speed limiting factor is the `read_xram()`s from the USB stick.  After the semester is over, I will try to polish it up a bit and put it on github.  Just for grins, the code for the scroll loop is below. It assumes `.addr0` and `.addr1` are set to the destination and source XRAM locations, and that the `.steps` are set to 1.

Looking at the code generated by cc65 can be very instructive.  Find the `.obj.s` files way down in the `build` directory.

As fun as hand-optimizing assembly is, I think an XRAM-to-XRAM `memcpy()` in the API would be cool (and way way faster).   It's on my list along with a bunch of other things.

```
        LDY #rows_to_scroll
aslp0:  LDX #width_bytes
aslp1:  LDA RIA_RW1
        STA RIA_RW0
        DEX
        BNE aslp1
        DEY
        BNE aslp0
        rts
```

**Reply by rumbledethumps** [reply]
> Looking at the code generated by cc65 can be very instructive. Find the `.obj.s` files way down in the `build` directory.

This and reading the cc65 docs. Trying to learn assembly RIA operations before you know how to use the cc65 compiler is like trying to run before you can walk. RIA operations are 100% based on cc65 internals. If you know how cc65 works then you will understand RIA operations.
